tenantcostclient: tenant-side throttling

RaduBerinde opened the pull request:
#### tenantcostserver: start with 10M RUs by default

This change modifies the default initial values for the token bucket -
relevant when crdb_internal.update_tenant_resource_limits was not
called. Specifically, we now start with 10M RUs by default. This is
intended to prevent unit tests from being throttled.

We also decrease the default rate, so it is in line with the planned
refill rate.

Release note: None

Release justification: multi-tenant-only code.

#### util: add support for tickers in TimeSource and ManualTime

This change will allow tests to manually control tickers.

Release justification: new library functionality, no changes to
existing production code.

Release note: None

#### tenantcostclient: report consumption when requests complete

Consumption is now reported only when requests complete. This will
allow us to avoid charging for failed requests (at least for certain
error conditions).

Release justification: multi-tenant-only code.

Release note: None

#### tenantcostclient: tenant-side throttling

This commit implements the bulk of the tenant-side throttling library.

High level overview of how it works (more details are in the RFC):
 - we start with an initial number of RUs (1000) that we can use
   immediately; in the meantime, we make an initial request to the
   bucket.

 - we request the tokens we estimate we'll need over the next 10s,
   according to an exponentially-moving average of recent consumption.
   We also factor in any token bucket debt or waiting requests.

 - if we are granted tokens immediately, we request more tokens when
   we have less than 10% of the granted tokens (i.e. 1s worth of
   usage) remaining.

 - if we are granted tokens over a period of time (i.e. a rate), we
   request more when we get within 1s of the "deadline" (unless we
   accumulate unused tokens (more than 10% of last granted).

Other than the datadriven test, I did a manual end-to-end test with a
KV workload and verified that we can throttle and un-throttle the
workload using `crdb_internal.update_tenant_resource_limits`.

The biggest things not yet done:

 - reasonable "backup" behavior when we can't perform token bucket
   requests (e.g. `tenant_usage` table is unavailable); I have a plan
   to return a backup rate from the server that would allow reasonable
   behavior for both free tenants and tenants with lots of paid RUs.

 - an end-to-end roachtest;

 - we account for CPU usage once per second, and take out the RUs from
   the buckets immediately. This has the potential of blocking out
   everything for a fraction of each second. This could be improved by
   increasing the sampling rate or implementing a debt-repayment
   mechanism where the debt is repaid continuously over the next
   second rather than instantly.

Release note: None

Release justification: multi-tenant-only code.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69895)
<!-- Reviewable:end -->

andy-kimball reviewed (commented):
> Reviewed 4 of 4 files at r1, 3 of 3 files at r2, 6 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69895)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andy-kimball, and @RaduBerinde)

___
*[pkg/ccl/multitenantccl/tenantcostclient/limiter.go, line 38 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1nl_oBSS4wqRJTeH3:-Mj1nl_oBSS4wqRJTeH4:b4pcqvs) ([raw file](https://github.com/cockroachdb/cockroach/blob/697eace7618901d0cafc51551724f34b70b9e8cc/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L38)):*
> ```Go
> 
> const initialRate = 100
> const initialRUs = 1000
> ```

How do these values interact with the default ones set in the other commit? Do these get quickly overridden by those other ones? Does CC need to do an up-front call to initialize the token bucket, or do these defaults mean it can afford to wait 10-20 seconds for its first collection and adjustment?
___
*[pkg/ccl/multitenantccl/tenantcostclient/limiter.go, line 72 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1hV5x6V1nnv_f2S0k:-Mj1hV5x6V1nnv_f2S0l:bu4zktf) ([raw file](https://github.com/cockroachdb/cockroach/blob/697eace7618901d0cafc51551724f34b70b9e8cc/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L72)):*
> ```Go
> }
> 
> // Adjust adds or removes tokens from the bucket. Tokens are added when we
> ```

NIT: AdjustTokens
___
*[pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go, line 185 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1pvejAvM3f3oq_g6x:-Mj1pvejAvM3f3oq_g6y:bb3zhq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/697eace7618901d0cafc51551724f34b70b9e8cc/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L185)):*
> ```Go
> 	c.run.cpuSecs = c.cpuSecsFn(ctx)
> 	c.run.lastRequestTime = now
> 	// Set up the initial avgRUPerSec so that the first request is for initialRUs.
> ```

Just to verify: when we spin up a new tenant cluster, have we set things up so that it can _immediately_ use 10K's of RUs without needing to wait for a round-trip to the tenant cost server? I want to make sure our new cluster creation experience doesn't have any unnecessary latency that's easy for us to remove (say by using different defaults).
___
*[pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go, line 218 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1vUUX8IxPk90vPbA9:-Mj1vUUX8IxPk90vPbAA:b7vsulo) ([raw file](https://github.com/cockroachdb/cockroach/blob/697eace7618901d0cafc51551724f34b70b9e8cc/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L218)):*
> ```Go
> 	c.run.consumption = newConsumption
> 
> 	// TODO(radu): figure out how to "smooth out" this debt over a longer period
> ```

Can we do something simple like sample CPU more often (say 4x a second), but still report less frequently? I don't have a good sense of how expensive it is to sample the CPU and go through the logic of subtracting it from the bucket, though.
___
*[pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go, line 224 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1uQPE3-dL2ML2zRki:-Mj1uQPE3-dL2ML2zRkj:b-68qyxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/697eace7618901d0cafc51551724f34b70b9e8cc/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L224)):*
> ```Go
> 
> func (c *tenantSideCostController) updateAvgRUPerSec() {
> 	c.run.avgRUPerSec = 0.5 * (c.run.avgRUPerSec + (c.run.consumption.RU - c.run.avgRUPerSecLastRU))
> ```

I'm confused by this formula. Doesn't it need some time component that measures how much time elapsed between sampling the `c.run.consumption.RU` value and the `c.run.avgRUPerSecLastRU` value? As it is, it looks like we're trying to add RU/s to RU, where the units don't match. Maybe this mostly "works" because you currently have the sampling period set to 1 second, but that seems fragile.
___
*[pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go, line 273 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1wUvY8JV17qekr5PO:-Mj1wUvY8JV17qekr5PP:b2bus6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/697eace7618901d0cafc51551724f34b70b9e8cc/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L273)):*
> ```Go
> 	req := roachpb.TokenBucketRequest{
> 		TenantID: c.tenantID.ToUint64(),
> 		// TODO(radu): populate instance ID.
> ```

What will happen if we aren't reporting the correct instance ID? Is this a critical "must have" TODO that we need to get done for the release?
___
*[pkg/ccl/multitenantccl/tenantcostclient/token_bucket.go, line 18 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1h4vz8ub_S7Dez_Zr:-Mj1h4vz8ub_S7Dez_Zs:b-ireha6) ([raw file](https://github.com/cockroachdb/cockroach/blob/697eace7618901d0cafc51551724f34b70b9e8cc/pkg/ccl/multitenantccl/tenantcostclient/token_bucket.go#L18)):*
> ```Go
> 
> // tokenBucket implements a token bucket. It is a more specialized
> // form of quotapool.TokenBucket.
> ```

Can you give more details in the comment of how it is more specialized?
___
*[pkg/ccl/multitenantccl/tenantcostclient/token_bucket.go, line 23 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1cpO5-s7aF-cMH-UM:-Mj1cpO6DKJvT6spv78k:bqbh2rv) ([raw file](https://github.com/cockroachdb/cockroach/blob/697eace7618901d0cafc51551724f34b70b9e8cc/pkg/ccl/multitenantccl/tenantcostclient/token_bucket.go#L23)):*
> ```Go
> 
> 	// When the available amount goes below the notifyThreshold and the time is
> 	// after notifyStartTime, we do a (non-blocking) send on the channel.
> ```

NIT on a NIT: the code below indicates it's `>= notifyStartTime` rather than `> notifyStartTime`.
___
*[pkg/ccl/multitenantccl/tenantcostclient/token_bucket.go, line 140 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1e8dSB9Xrut2CzQbl:-Mj1e8dSB9Xrut2CzQbm:bs3gqjv) ([raw file](https://github.com/cockroachdb/cockroach/blob/697eace7618901d0cafc51551724f34b70b9e8cc/pkg/ccl/multitenantccl/tenantcostclient/token_bucket.go#L140)):*
> ```Go
> 	timeSeconds := float64((amount - tb.available) / tb.rate)
> 	// Cap the number of seconds to avoid overflow; we want to tolerate even a
> 	// rate of 0 (in which case we are really waiting for a token adjustment).
> ```

If we're waiting for a token adjustment, should we really wait for 16 minutes until we call back? Could this ever happen before the server has initialized the token bucket and cause long stalls?
___
*[pkg/util/timeutil/time_source.go, line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1Gyv9C4lT0QOcbozN:-Mj1Gyv9C4lT0QOcbozO:bcm216n) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4bfae128c525f3582654eb8b7d9abd4efda4689/pkg/util/timeutil/time_source.go#L41)):*
> ```Go
> }
> 
> // TickerI is an interface wrapping Ticker.
> ```

How come `TickerI` can't be trivially simulated with `TimerI` by calling `Reset` upon receiving the `Ch` channel notification so that it continually calls back?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-:-Mj4kLUMBgcEIJMzOVO4:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @andy-kimball)

___
*[pkg/ccl/multitenantccl/tenantcostclient/limiter.go, line 38 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1nl_oBSS4wqRJTeH3:-Mj4_L9s14ZmDgPmkLVp:b-nz3zhl) ([raw file](https://github.com/cockroachdb/cockroach/blob/697eace7618901d0cafc51551724f34b70b9e8cc/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L38)):*
Adding this comment:
```
// Initial settings for the local token bucket. They are used only until the
// first TokenBucket request returns. We allow immediate use of the initial RUs
// (we essentially borrow them and pay them back in the first TokenBucket
// request). The intention is to avoid any throttling during start-up in normal
// circumstances.
```

> Does CC need to do an up-front call to initialize the token bucket, or do these defaults mean it can afford to wait 10-20 seconds for its first collection and adjustment?

The proposed global settings are consistent with both a free and paid user for some initial period: no throttling until we eat through the 10M. So yes, the call to configure the bucket can come quite a bit later.
___
*[pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go, line 185 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1pvejAvM3f3oq_g6x:-Mj4_aGgC76TUsLLwr5X:b-9a9fvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/697eace7618901d0cafc51551724f34b70b9e8cc/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L185)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Just to verify: when we spin up a new tenant cluster, have we set things up so that it can _immediately_ use 10K's of RUs without needing to wait for a round-trip to the tenant cost server? I want to make sure our new cluster creation experience doesn't have any unnecessary latency that's easy for us to remove (say by using different defaults).
</blockquote></details>

We can immediately use the `initialRUs`. I set them at 1K (thinking that we should normally get the token bucket request through in at most a few hundred ms). Do you think 10K would be a better idea?
___
*[pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go, line 218 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1vUUX8IxPk90vPbA9:-Mj4d5vp00XiAMKi-Wpv:bgesfgq) ([raw file](https://github.com/cockroachdb/cockroach/blob/697eace7618901d0cafc51551724f34b70b9e8cc/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L218)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Can we do something simple like sample CPU more often (say 4x a second), but still report less frequently? I don't have a good sense of how expensive it is to sample the CPU and go through the logic of subtracting it from the bucket, though.
</blockquote></details>

That's definitely one possible fix (though maybe partial). I want to stay away from that high idle usage problem, but a few times per second would probably be ok.

The other solution I was thinking of was to adjust the rate on the local bucket to pay for the debt over the next 1s.
___
*[pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go, line 224 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1uQPE3-dL2ML2zRki:-Mj4eC3t7Q1Zme2D8jBZ:bo4vzdd) ([raw file](https://github.com/cockroachdb/cockroach/blob/697eace7618901d0cafc51551724f34b70b9e8cc/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L224)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'm confused by this formula. Doesn't it need some time component that measures how much time elapsed between sampling the `c.run.consumption.RU` value and the `c.run.avgRUPerSecLastRU` value? As it is, it looks like we're trying to add RU/s to RU, where the units don't match. Maybe this mostly "works" because you currently have the sampling period set to 1 second, but that seems fragile.
</blockquote></details>

Yeah, that was the idea. This is only called once per "tick" (and not for other wakeups). I made it a constant and added more clarification, but I kept this simple scheme.
___
*[pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go, line 273 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1wUvY8JV17qekr5PO:-Mj4eMC6EoteGW4Krmzc:b6pjioh) ([raw file](https://github.com/cockroachdb/cockroach/blob/697eace7618901d0cafc51551724f34b70b9e8cc/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L273)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

What will happen if we aren't reporting the correct instance ID? Is this a critical "must have" TODO that we need to get done for the release?
</blockquote></details>

The only problem is when we throttle and have multiple instances - we'd give out the full rate to all the instances. Over time the debt mechanism in the global bucket would account for that and bring their rates down to pay for the debt.

It will be totally fine if we expect that (for the most part) we run only one instance for throttled tenants.

I do want to get the instance ID stuff in soon though, to avoid upgrade headaches later.
___
*[pkg/ccl/multitenantccl/tenantcostclient/token_bucket.go, line 18 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1h4vz8ub_S7Dez_Zr:-Mj4fMGdC4xZ98ltPC_R:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/697eace7618901d0cafc51551724f34b70b9e8cc/pkg/ccl/multitenantccl/tenantcostclient/token_bucket.go#L18)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Can you give more details in the comment of how it is more specialized?
</blockquote></details>

Done.
___
*[pkg/ccl/multitenantccl/tenantcostclient/token_bucket.go, line 140 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1e8dSB9Xrut2CzQbl:-Mj4gE276-J4tQs84Lqx:b-40wdgj) ([raw file](https://github.com/cockroachdb/cockroach/blob/697eace7618901d0cafc51551724f34b70b9e8cc/pkg/ccl/multitenantccl/tenantcostclient/token_bucket.go#L140)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

If we're waiting for a token adjustment, should we really wait for 16 minutes until we call back? Could this ever happen before the server has initialized the token bucket and cause long stalls?
</blockquote></details>

The wait time here is how long we need to wait until a retry of the `TryToFulfill` call would succeed if nothing changes (it can be smaller too, just causing more wakeups). If this is very long, we are essentially waiting for a change (a `Reconfigure` or an `AdjustToken` call); these changes independently trigger a re-evaluation of `TryToFulfill`.

Before we initialize the bucket, we have 100 refill rate and things would be allowed to go through at that rate. The danger here is when we're running unthrottled (where the rate is 0) and we stop being able to communicate with the global token bucket. This is the TODO I mentioned in the commit - I will work on having a "backup" rate that we would fall back to.
___
*[pkg/util/timeutil/time_source.go, line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1Gyv9C4lT0QOcbozN:-Mj4iHJBCNA3IP-b2dS3:b-a4irdn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4bfae128c525f3582654eb8b7d9abd4efda4689/pkg/util/timeutil/time_source.go#L41)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

How come `TickerI` can't be trivially simulated with `TimerI` by calling `Reset` upon receiving the `Ch` channel notification so that it continually calls back?
</blockquote></details>

Good question. I was using the ticker API and just wanted to keep using that. Thinking about it, with a manual `Reset()` each time we would need to account for any scheduling delays if we want to keep the "tick" consistent (we want the period to be 1s, not `1s + current scheduling delay`)


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-:-Mj4qap_1SvrKYbvJkfV:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @andy-kimball, and @RaduBerinde)

___
*[pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go, line 185 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1pvejAvM3f3oq_g6x:-Mj4pQvcBfecUZVgPArl:bk1kg7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/697eace7618901d0cafc51551724f34b70b9e8cc/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L185)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We can immediately use the `initialRUs`. I set them at 1K (thinking that we should normally get the token bucket request through in at most a few hundred ms). Do you think 10K would be a better idea?
</blockquote></details>

Yeah, I think 10K would be better.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69895)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @andy-kimball)

___
*[pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go, line 185 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1pvejAvM3f3oq_g6x:-Mj5RCSRCoVpZgLhkxk6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/697eace7618901d0cafc51551724f34b70b9e8cc/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L185)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Yeah, I think 10K would be better.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner commented on pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go:
nit: `AvgRu` vs `AvgRU`

ajwerner commented on pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go:
method name doesn't match the comment.

ajwerner commented on pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go:
some commentary about how `requestNeedsRetry` interacts with `requestInProgress` might be handy. Very nit-picky but I'd put initialRequestCompleted above `requestInProgress` because those other two interact with each othe.

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69895)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @andy-kimball)

___
*[pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go, line 57 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1827U-bXmoo3KM6ev-r7-57:-Mj6tfZ6-sA29Z_9QjUF:bslmwc1) ([raw file](https://github.com/cockroachdb/cockroach/blob/650a73f2f852be68616d3e27c5da2255b077d639/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L57)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I think this approach of changing the rate of decay based on the rate of the loop is fine, but I wanted to share one of my favorite posts because I think it's really cool. I am not at all suggesting you adopt it, and, to be honest, I've never used it, but it does make me very happy. http://tdunning.blogspot.com/2011/03/exponential-weighted-averages-with.html 

The same author has a different post that points out that when dealing with rates there can be problems, but he also concludes it probably doesn't matter. http://tdunning.blogspot.com/2011/03/exponentially-weighted-averaging-for.html
</blockquote></details>

Heh, that's a neat trick, thanks for sharing! I was naively thinking that you can still use the mixing formula (with `mu = e^-Δt`), but I see how it doesn't work when you have samples that are very close together.

It's worth noting that the mixing trick isn't quite equivalent to the "right" formula even for equally spaced samples (they are only equivalent "in the limit", i.e. after many measurements)
___
*[pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go, line 156 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1827U-bXmoo3KM6ev-r7-156:-Mj7-Iww2S_rKZX7i-fD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/650a73f2f852be68616d3e27c5da2255b077d639/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L156)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

some commentary about how `requestNeedsRetry` interacts with `requestInProgress` might be handy. Very nit-picky but I'd put initialRequestCompleted above `requestInProgress` because those other two interact with each othe.
</blockquote></details>

Done.
___
*[pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go, line 243 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-Mj1827U-bXmoo3KM6ev-r7-243:-Mj7-Wya62iH5kIat2-H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/650a73f2f852be68616d3e27c5da2255b077d639/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L243)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

method name doesn't match the comment.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

RaduBerinde commented:
bors r-

I was working on a follow-up change and I think I found a bug.

RaduBerinde commented:
Ok, I fixed some things. I kept the new changes in a separate commit but I will squash it before merging.

#### fixes and improvements

This change fixes some bugs due to race conditions and introduces more
primitives for writing tests.

We now use a timer to set up the "low RU" notification instead of
implementing the start time in the token bucket - this was buggy in
the case where we had more than the threshold RUs in the bucket but an
even larger request was waiting.

We also fix a bug when the "low RU" notification happens while we are
doing a token bucket request which doesn't request new RUs. We were
effectively losing the notification, which could lead to blocking
until the next reporting period.

We introduce an instrumentation interface which allows the test to
wait for specific internal events. We also add the `timers` directive
(borrowed from `tenantrate`) which can be used to wait for the quota
pool timers to be set up.

Finally, we add an `OnStartLocked` callback to `quotapool` which we
use to keep track of RUs that are needed for waiting request. The
existing `OnStart` callback was insufficient because there was a
window where the waiting RUs are not visible (and if we have a single
request, it would cause us to not request more RUs).

andy-kimball reviewed (commented):
> Reviewed 2 of 6 files at r4, 1 of 3 files at r5, 2 of 2 files at r6, 1 of 2 files at r7, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69895)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @andy-kimball, and @RaduBerinde)

___
*[pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go, line 194 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGdhEp9cClQ8WckyUc:-MjGdhEp9cClQ8WckyUd:b-jctvq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3922aa69145fad0775d3cb6e07d34fe3ca683370/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L194)):*
> ```Go
> 		setupNotificationTimer     timeutil.TimerI
> 		setupNotificationCh        <-chan time.Time
> 		setupNotificationThreshold tenantcostmodel.RU
> ```

I don't see where `setupNotificationThreshold` is being used, just where it's being set.
___
*[pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go, line 426 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGcL3P659GNh8PFlLJ:-MjGcL3P659GNh8PFlLK:b-2sed78) ([raw file](https://github.com/cockroachdb/cockroach/blob/3922aa69145fad0775d3cb6e07d34fe3ca683370/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L426)):*
> ```Go
> 		}
> 
> 		c.run.setupNotificationTimer = c.timeSource.NewTimer()
> ```

How come it's necessary to set up a new timer? Isn't `maybeNotify` already being called frequently enough without a timer? Couldn't the logic in there be changed to see if the remaining time is low in addition to checking the remaining RUs to see if they've dropped below `NotifyThreshold`?
___
*[pkg/util/quotapool/config.go, line 132 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGf1McEwmTMl_ELrWz:-MjGf1McEwmTMl_ELrX-:b-hmv5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/3922aa69145fad0775d3cb6e07d34fe3ca683370/pkg/util/quotapool/config.go#L132)):*
> ```Go
> 	onAcquisition            AcquisitionFunc
> 	onSlowAcquisition        SlowAcquisitionFunc
> 	onWaitStartLocked        OnWaitStartFunc
> ```

There's no other good way to accomplish what you need to do here? I always try to avoid firing events when I've acquired a lock, as it inevitably seems to lead to re-entrancy deadlocks, callbacks that make network calls under the scope of the lock, etc. I think setting this precedent will make the `AbstractQuotaPool` easier to accidentally abuse.

CC @ajwerner for his thoughts/suggestions


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69895)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @RaduBerinde)

___
*[pkg/ccl/multitenantccl/tenantcostclient/limiter.go, line 34 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGdkFt2nSde4iLmiPq:-MjGdkFt2nSde4iLmiPr:b-25jamo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3922aa69145fad0775d3cb6e07d34fe3ca683370/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L34)):*
> ```Go
> 	// Total (rounded) RU needed for all currently waiting requests (or requests
> 	// that are in the process of being fulfilled).
> 	waitingRU int64
> ```

note the use of atomics
___
*[pkg/ccl/multitenantccl/tenantcostclient/testdata/throttling, line 27 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGfwIEBTMkDtUG8xef:-MjGfwIEBTMkDtUG8xeg:b-2hj2cp) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaf757cd83447728d61d2e1359477e7555a718cd/pkg/ccl/multitenantccl/tenantcostclient/testdata/throttling#L27)):*
> ```
> 
> # Set up throttling at 1000 RU/s.
> throttle 
> ```

nit: trailing spaces
___
*[pkg/ccl/multitenantccl/tenantcostserver/testdata/requests, line 10 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGg7PJ3PmWmMJlqL03:-MjGg7PJ3PmWmMJlqL04:b-2hj2cp) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaf757cd83447728d61d2e1359477e7555a718cd/pkg/ccl/multitenantccl/tenantcostserver/testdata/requests#L10)):*
> ```
> # First drain all the initially available tokens.
> token-bucket-request tenant=5
> instance_id: 1 
> ```

nit: trailing spaces
___
*[pkg/ccl/multitenantccl/tenantcostserver/testdata/requests, line 16 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGgA9M9bqCwqX5ieja:-MjGgA9M9bqCwqX5iejb:b-2hj2cp) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaf757cd83447728d61d2e1359477e7555a718cd/pkg/ccl/multitenantccl/tenantcostserver/testdata/requests#L16)):*
> ```
> 
> token-bucket-request tenant=5
> instance_id: 1 
> ```

nit: trailing spaces
___
*[pkg/ccl/multitenantccl/tenantcostserver/testdata/requests, line 49 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGgBZ1A7qh0qSyBtrx:-MjGgBZ24wsFEvUw-lvv:b-2hj2cp) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaf757cd83447728d61d2e1359477e7555a718cd/pkg/ccl/multitenantccl/tenantcostserver/testdata/requests#L49)):*
> ```
> 
> token-bucket-request tenant=5
> instance_id: 2 
> ```

nit: trailing spaces
___
*[pkg/util/quotapool/config.go, line 132 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGf1McEwmTMl_ELrWz:-MjGgH9GDDacw0RtnC2S:bxjd2iu) ([raw file](https://github.com/cockroachdb/cockroach/blob/3922aa69145fad0775d3cb6e07d34fe3ca683370/pkg/util/quotapool/config.go#L132)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

There's no other good way to accomplish what you need to do here? I always try to avoid firing events when I've acquired a lock, as it inevitably seems to lead to re-entrancy deadlocks, callbacks that make network calls under the scope of the lock, etc. I think setting this precedent will make the `AbstractQuotaPool` easier to accidentally abuse.

CC @ajwerner for his thoughts/suggestions
</blockquote></details>

We spent a while going back and forth here. The issues were largely about testing. How do you ensure that after you've sent the notification that the bookkeeping is up-to-date? 

We played with some other ideas that ended up feeling heavier. At least this has the word `Locked` on it which should dissuade folks from doing anything too crazy. @raduberinde can you remind me why updating `waitingRUs` and notifying in the existing `onWaitStart` callback under a separate mutex was problematic? Was it because it doesn't serialize with potential updates due to request callbacks in the meantime? 


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/util/quotapool/config.go, line 132 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGf1McEwmTMl_ELrWz:-MjGvWk93LFIfFRPqimN:bag35y3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3922aa69145fad0775d3cb6e07d34fe3ca683370/pkg/util/quotapool/config.go#L132)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

We spent a while going back and forth here. The issues were largely about testing. How do you ensure that after you've sent the notification that the bookkeeping is up-to-date? 

We played with some other ideas that ended up feeling heavier. At least this has the word `Locked` on it which should dissuade folks from doing anything too crazy. @raduberinde can you remind me why updating `waitingRUs` and notifying in the existing `onWaitStart` callback under a separate mutex was problematic? Was it because it doesn't serialize with potential updates due to request callbacks in the meantime? 
</blockquote></details>

This was not a testing issue, it's a real bug. The race is this: we get a large request; it triggers a notification; we handle the notification and calculate how many RUs we need before the callback runs. At this point we don't "see" the RUs of the large request. It looks like we don't need more RUs so we don't ask for more. The situation will only be remedied in 10 or 40s when we do a periodic request.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69895)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @RaduBerinde)

___
*[pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go, line 194 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGdhEp9cClQ8WckyUc:-MjGuSlv0B0BZ0e3mO-Y:b-s8hfqc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3922aa69145fad0775d3cb6e07d34fe3ca683370/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L194)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I don't see where `setupNotificationThreshold` is being used, just where it's being set.
</blockquote></details>

there's a SetupNotification call that uses it
___
*[pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go, line 426 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGcL3P659GNh8PFlLJ:-MjGuXCO4tYwegE32mx8:b-wbniag) ([raw file](https://github.com/cockroachdb/cockroach/blob/3922aa69145fad0775d3cb6e07d34fe3ca683370/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L426)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

How come it's necessary to set up a new timer? Isn't `maybeNotify` already being called frequently enough without a timer? Couldn't the logic in there be changed to see if the remaining time is low in addition to checking the remaining RUs to see if they've dropped below `NotifyThreshold`?
</blockquote></details>

If we have more RUs than the threshold but we have an even larger request already waiting, we won't get a notification. We could check periodically but that seems worse than a timer.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69895)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @RaduBerinde)

___
*[pkg/util/quotapool/config.go, line 132 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGf1McEwmTMl_ELrWz:-MjGx6rJ9UzRgY3oCmrW:bfx8cvi) ([raw file](https://github.com/cockroachdb/cockroach/blob/3922aa69145fad0775d3cb6e07d34fe3ca683370/pkg/util/quotapool/config.go#L132)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This was not a testing issue, it's a real bug. The race is this: we get a large request; it triggers a notification; we handle the notification and calculate how many RUs we need before the callback runs. At this point we don't "see" the RUs of the large request. It looks like we don't need more RUs so we don't ask for more. The situation will only be remedied in 10 or 40s when we do a periodic request.
</blockquote></details>

The question I'm asking is what if we didn't trigger the notification until we were in the callback?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69895)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @RaduBerinde)

___
*[pkg/util/quotapool/config.go, line 132 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGf1McEwmTMl_ELrWz:-MjGxcVL0zR4ASdIfvjA:bxpd4b0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3922aa69145fad0775d3cb6e07d34fe3ca683370/pkg/util/quotapool/config.go#L132)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

The question I'm asking is what if we didn't trigger the notification until we were in the callback?
</blockquote></details>

Is the problem there that we want to also notify again sometimes while we're waiting?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69895)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @RaduBerinde)

___
*[pkg/util/quotapool/config.go, line 132 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGf1McEwmTMl_ELrWz:-MjGxsCh7ll0rHcvvrNx:bm1q6ti) ([raw file](https://github.com/cockroachdb/cockroach/blob/3922aa69145fad0775d3cb6e07d34fe3ca683370/pkg/util/quotapool/config.go#L132)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Is the problem there that we want to also notify again sometimes while we're waiting?
</blockquote></details>

Would it be better if we just plumbed the boolean of whether or not this is the first time we're processing the request to the `Acquire` interface? 


<!-- Sent from Reviewable.io -->

andy-kimball commented:
___
*[pkg/util/quotapool/config.go, line 132 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGf1McEwmTMl_ELrWz:-MjH3hg7C6TlKd7we-dm:b-och00h) ([raw file](https://github.com/cockroachdb/cockroach/blob/3922aa69145fad0775d3cb6e07d34fe3ca683370/pkg/util/quotapool/config.go#L132)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Would it be better if we just plumbed the boolean of whether or not this is the first time we're processing the request to the `Acquire` interface? 
</blockquote></details>

If you two agree that this is the right solution, then I'm OK with it. It triggers my spidey sense, but I also don't have a suggestion as to how to do this differently - there are very subtle issues at play that would take many hours to fully understand.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/util/quotapool/config.go, line 132 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGf1McEwmTMl_ELrWz:-MjH2cb57737jRDPLE-m:b-6p4ihv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3922aa69145fad0775d3cb6e07d34fe3ca683370/pkg/util/quotapool/config.go#L132)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

If you two agree that this is the right solution, then I'm OK with it. It triggers my spidey sense, but I also don't have a suggestion as to how to do this differently - there are very subtle issues at play that would take many hours to fully understand.
</blockquote></details>

We also notify when the RUs get low even if we're not blocking requests.

Sending the notification in the callback would fix this case, but we'd need to acquire the lock to check if we need to send it. Calling quotapool.Update from this callback feels much worse.

I don't immediately see a cleaner solution even with plumbing the boolean.

As an aside - we (Andrew and I) discussed this over slack earlier today and this was the best we could come up with. I think we all agree something is amiss here but the solution continues to elude me.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-:-MjH7iH-9ckbX6VKOGJT:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @andy-kimball, and @RaduBerinde)

___
*[pkg/util/quotapool/config.go, line 132 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGf1McEwmTMl_ELrWz:-MjH7Clr8ujV0Cxc1DXD:bumlcl) ([raw file](https://github.com/cockroachdb/cockroach/blob/3922aa69145fad0775d3cb6e07d34fe3ca683370/pkg/util/quotapool/config.go#L132)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We also notify when the RUs get low even if we're not blocking requests.

Sending the notification in the callback would fix this case, but we'd need to acquire the lock to check if we need to send it. Calling quotapool.Update from this callback feels much worse.

I don't immediately see a cleaner solution even with plumbing the boolean.

As an aside - we (Andrew and I) discussed this over slack earlier today and this was the best we could come up with. I think we all agree something is amiss here but the solution continues to elude me.
</blockquote></details>

Make sure to add a big comment somewhere explaining the subtleties here and the reason why we're using `OnWaitStartLocked`. Also, put a warning note in the comment for `OnWaitStartLocked` to warn future developers that they need to be very careful in their callback method - no re-entrancy allowed and it should never block.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69895)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball and @RaduBerinde)

___
*[pkg/util/quotapool/config.go, line 132 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGf1McEwmTMl_ELrWz:-MjHByN1DRTPNXW4J54T:b-30yl1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/3922aa69145fad0775d3cb6e07d34fe3ca683370/pkg/util/quotapool/config.go#L132)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Make sure to add a big comment somewhere explaining the subtleties here and the reason why we're using `OnWaitStartLocked`. Also, put a warning note in the comment for `OnWaitStartLocked` to warn future developers that they need to be very careful in their callback method - no re-entrancy allowed and it should never block.
</blockquote></details>

How do y'all feel about this change to the API instead. I think it cleans up the grossness pretty well.  https://github.com/cockroachdb/cockroach/commit/17b08ede743716c138ac656db78383fa12300388


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69895)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball and @RaduBerinde)

___
*[pkg/util/quotapool/config.go, line 132 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-MjGf1McEwmTMl_ELrWz:-MjHDkmD5SSHn2iAru8C:bcp8gvn) ([raw file](https://github.com/cockroachdb/cockroach/blob/3922aa69145fad0775d3cb6e07d34fe3ca683370/pkg/util/quotapool/config.go#L132)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

How do y'all feel about this change to the API instead. I think it cleans up the grossness pretty well.  https://github.com/cockroachdb/cockroach/commit/17b08ede743716c138ac656db78383fa12300388
</blockquote></details>

Err: https://github.com/cockroachdb/cockroach/commit/2e40f355aa952b595defd015df72b0980c486e2a actually adopts the new API properly. 


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
^ Well, now you're only counting the waiting RUs for the head of the queue. I want to take all waiting requests into account when requesting more RUs.

One way to avoid the "locked callback" would be to keep this flag in the Request, and increment `waiting` either inside Acquire or in the regular start callback, whichever sees the request first.

RaduBerinde commented:
> One way to avoid the "locked callback" would be to keep this flag in the Request, and increment waiting either inside Acquire or in the regular start callback, whichever sees the request first.

Like this: https://github.com/RaduBerinde/cockroach/commit/5153baa53c58cecccb715bc0e45d0ddb8fd9bb0b

A tad messier in this code but we don't pollute AbstractPool with the new callback.

ajwerner reviewed (commented):
ah yeah... forgot the whole thing with the fast path is we don't call it if there's something queued. I think this is better. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-:-MjHkpq7DMX3MSN9yUM1:bl8saaw)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @RaduBerinde)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
Updated, made the update in a separate commit. I also decided to decrement `waitingRU` in Acquire (under the lock), similar to how your change did it. This is an improvement in that we never double-count the RUs (i.e. observe a reduced amount in the bucket AND see them as `waitingRU`s).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-:-MjHoQ-PAnQnZZS7TezP:bxw6nqw)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball)



<!-- Sent from Reviewable.io -->

andy-kimball commented:
Glad you two found a design that kept the interface nicer.





<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69895#-:-MjHzDFj-4XsxesBks7O:b-jbsf8z)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball)



<!-- Sent from Reviewable.io -->

RaduBerinde committed:
tenantcostserver: start with 10M RUs by default

This change modifies the default initial values for the token bucket -
relevant when crdb_internal.update_tenant_resource_limits was not
called. Specifically, we now start with 10M RUs by default. This is
intended to prevent unit tests from being throttled.

We also decrease the default rate, so it is in line with the planned
refill rate.

Release note: None

Release justification: multi-tenant-only code.

RaduBerinde committed:
tenantcostclient: report consumption when requests complete

Consumption is now reported only when requests complete. This will
allow us to avoid charging for failed requests (at least for certain
error conditions).

Release justification: multi-tenant-only code.

Release note: None

RaduBerinde committed:
tenantcostclient: tenant-side throttling

This commit implements the bulk of the tenant-side throttling library.

High level overview of how it works (more details are in the RFC):
 - we start with an initial number of RUs (1000) that we can use
   immediately; in the meantime, we make an initial request to the
   bucket.

 - we request the tokens we estimate we'll need over the next 10s,
   according to an exponentially-moving average of recent consumption.
   We also factor in any token bucket debt or waiting requests.

 - if we are granted tokens immediately, we request more tokens when
   we have less than 10% of the granted tokens (i.e. 1s worth of
   usage) remaining.

 - if we are granted tokens over a period of time (i.e. a rate), we
   request more when we get within 1s of the "deadline" (unless we
   accumulate unused tokens (more than 10% of last granted).

Other than the datadriven test, I did a manual end-to-end test with a
KV workload and verified that we can throttle and un-throttle the
workload using `crdb_internal.update_tenant_resource_limits`.

The biggest things not yet done:

 - reasonable "backup" behavior when we can't perform token bucket
   requests (e.g. `tenant_usage` table is unavailable); I have a plan
   to return a backup rate from the server that would allow reasonable
   behavior for both free tenants and tenants with lots of paid RUs.

 - an end-to-end roachtest;

 - we account for CPU usage once per second, and take out the RUs from
   the buckets immediately. This has the potential of blocking out
   everything for a fraction of each second. This could be improved by
   increasing the sampling rate or implementing a debt-repayment
   mechanism where the debt is repaid continuously over the next
   second rather than instantly.

Release note: None

Release justification: multi-tenant-only code.

RaduBerinde committed:
util: add support for tickers in TimeSource and ManualTime

This change will allow tests to manually control tickers.

Release justification: new library functionality, no changes to
existing production code.

Release note: None

RaduBerinde commented:
TFTRs!

CI failure is an unrelated flake.

bors r+

RaduBerinde commented:
Argh, that last change I made (to remove the "locked on wait" callback) now causes a test flake.. (we either count or not count a second waiting request). I'll try to figure out a way to fix.

RaduBerinde commented:
Ah, false alarm, never mind, I messed up (I had some changes in the branch).
